### PR TITLE
Optimize torch.eye(d) to avoid parallel_for overhead for small n

### LINKED ISSUE #48251: torch.eye(d) is slow and hogs cpu for d >= 182
## 🐛 Bug

I stumbled upon excessive CPU usage for my training code running on GPU. After some investigations I found the culprit.
It basically was
```
x = torch.eye(256).to('cuda') 
```

### This is quick and loads single CPU core.
```
%%timeit
    torch.eye(181)
6.43 µs ± 218 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
### This is 3 times slower and hogs 30 cores.
```
%%timeit
    torch.eye(182)
21.2 µs ± 84.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
## Environment

I test on server with 72 cores (hyper-threading is on)

```
% python3 collect_env.py

Collecting environment information...
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti

Nvidia driver version: 440.100
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.2
[pip3] pytorch-wavelets==1.2.2
[pip3] torch==1.6.0
[pip3] torchvision==0.4.2
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by gumeeee**
## Summary

Fixes #48251

`torch.eye(n)` is ~17x slower for n=182 than for n=181 due to
`parallel_for` thread-pool initialization overhead once
`n*n` crosses the TensorIterator parallelization threshold
(~32768 elements).

This shows up as a clear step function:
- `torch.eye(181)` ~ 1.9 us
- `torch.eye(182)` ~ 35 us  (17x jump)

The same jump repeats at every multiple where `n*n` crosses the
threshold, affecting any user who creates eye matrices in the
range of `182 <= n < 32768`.

## Approach

For the common square contiguous case, write the diagonal directly
via a flat-index loop, skipping the `parallel_for` dispatch entirely.
Non-square and non-contiguous cases fall back to the original
`parallel_for` path.

```cpp
if (n == m && result.is_contiguous()) {
  // Direct flat-index diagonal write
  scalar_t* result_data = result.data_ptr<scalar_t>();
  for (int64_t i = 0; i < sz; ++i) {
    result_data[i * (n + 1)] = 1;
  }
  return result;
}
// existing parallel_for path for non-square or non-contiguous
```

## Edge cases handled

- n == 0 or m == 0: early-return (empty matrix)
- Non-square (n != m): original path preserved
- Non-contiguous result: original path preserved
- All dtypes via AT_DISPATCH_V2 (bf16, f16, bool, floats, complex, f8)

## Test Plan

Added in `test/test_tensor_creation_ops.py`:
- `test_eye_no_threshold_regression`: correctness across the
  threshold for sizes 50, 100, 150, 181, 182, 200, 256
- `test_eye_non_square_unchanged`: (3,5), (5,3), (4,7), (7,4)
- `test_eye_zero_dim`: 0-dim edge case

Validated equivalence to `torch.eye` for:
- d in [1, 2, 5, 10, 50, 100, 181, 182, 256, 500, 1000]
- non-square pairs: (3,5), (5,3), (4,7), (7,4), (100,200)
- zero-dim: (0,5), (5,0), (0,0)
- dtypes: float32, float64, int32, int64, complex64, complex128

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Review by ngimel — COMMENTED** [review]
Looks like parallel dispatch won't ever be beneficial for this op, even if output is discontiguous?

**Comment by gumeeee** [comment]
You're right parallel_for isn't worth it here. Pure overhead for this op.

evidence:

1. eye_out_cuda uses as_strided + fill_(1) without parallel_for, same op, same result.
2. The ATen convention is sequential when numel < GRAIN_SIZE (32768). The current eye_out_cpu ignores this.
3. The diagonal is sz = min(n, m) memory-bound writes each thread does one write, no sharing, no contention. Thread-pool init (tens of µs) dominates any parallelism for small sz.

my plan: drop the second AT_DISPATCH_V2 and use a single flat-index loop (matches the CUDA path). I'll push the simplified version shortly.
